### PR TITLE
Configurable tags for prefixes and filter for storage class

### PR DESCRIPTION
**Pull request by cchantep**
Allow to configure the exporter to expose specific prefix tags and to filter for a specific storage class (e.g. exclude GLACIER objects)

### CODE REVIEW

File: s3_exporter.go [hunk 1]
```diff
@@ -26,45 +27,46 @@ var (
 	s3ListSuccess = prometheus.NewDesc(
 		prometheus.BuildFQName(namespace, "", "list_success"),
 		"If the ListObjects operation was a success",
-		[]string{"bucket", "prefix"}, nil,
+		[]string{"bucket", "prefix", "storage_class"}, nil,
```
Comment (cchantep) [inline s3_exporter.go:30]: New tag, `*` if no storage class configured

File: s3_exporter.go [hunk 2]
```diff
@@ -80,82 +82,142 @@ func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
...
-	var biggestObjectSize int64
-	var lastObjectSize int64
-
-	query := &s3.ListObjectsV2Input{
-		Bucket: &e.bucket,
-		Prefix: &e.prefix,
-	}
+	log.Infoln("Bucket: ", e.bucket)
+
+	for _, prefix := range e.prefixes {
```
Comment (cchantep) [inline s3_exporter.go:87]: Support filter multiple prefixes

File: s3_exporter.go [hunk 3]
```diff
@@ -80,82 +82,142 @@ func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
...
+		var lastModified time.Time
+		var numberOfObjects float64
+		var totalSize int64
+		var biggestObjectSize int64
+		var lastObjectSize int64
+
+		log.Infoln("Prefix: ", prefix)
+
+		labels := []string{e.bucket, prefix}
```
Comment (cchantep) [inline s3_exporter.go:96]: Prepare labels/tabs with at least bucket & prefix ...

File: s3_exporter.go [hunk 4]
```diff
@@ -80,82 +82,142 @@ func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
...
-	startList := time.Now()
-	for {
-		resp, err := e.svc.ListObjectsV2(query)
-		if err != nil {
-			log.Errorln(err)
-			ch <- prometheus.MustNewConstMetric(
-				s3ListSuccess, prometheus.GaugeValue, 0, e.bucket, e.prefix,
-			)
-			return
+		if e.storageclass != "" {
```
Comment (cchantep) [inline s3_exporter.go:98]: ... optionnally with storage class if configured

File: s3_exporter.go [hunk 5]
```diff
@@ -80,82 +82,142 @@ func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
...
+			if err != nil {
+				log.Errorln(err)
+				ch <- prometheus.MustNewConstMetric(
+					s3ListSuccess, prometheus.GaugeValue, 0, e.bucket, prefix,
+				)
+				return
+			}
+			for _, item := range resp.Contents {
+				if e.storageclass != "" &&
+					*item.StorageClass != e.storageclass {
```
Comment (cchantep) [inline s3_exporter.go:122]: Filter out non matching storage class

File: s3_exporter.go [hunk 6]
```diff
@@ -80,82 +82,142 @@ func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
...
+			s3BiggestSize, prometheus.GaugeValue, float64(biggestObjectSize), labels...,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			s3SumSize, prometheus.GaugeValue, float64(totalSize), labels...,
+		)
+	}
 }
 
-func probeHandler(w http.ResponseWriter, r *http.Request, svc s3iface.S3API) {
+func probeHandler(w http.ResponseWriter, r *http.Request, svc s3iface.S3API, cfgBucket string, cfgPrefixes string, cfgStorageClass string) {
```
Comment (cchantep) [inline s3_exporter.go:172]: Support from configuration

**Comment by cchantep** [comment]
up

**Comment by ribbybibby** [comment]
A couple of thoughts:
- Rather than using a parameter to filter based on storage class, what we could do is export metrics per-storage class for every storage class by default. Then users can filter based on storage class in PromQL without having to specifically set a storage class for each probe. This gives extra flexibility without an added configuration burden.
- Do we need an extra `prefixes` parameter? Can't you just define a separate target for each prefix?

**Comment by ribbybibby** [comment]
Also, this PR is adding a lot of different features:
- Adds storage class label
- Adds 'prefixes' parameter to scrape multiple prefixes in one probe
- Adds default values for parameters configured per-exporter

This makes it difficult to assess each change individually so I'd suggest separate PRs.

**Comment by cchantep** [comment]
> Rather than using a parameter to filter based on storage class, what we could do is export metrics per-storage class for every storage class by default. Then users can filter based on storage class in PromQL without having to specifically set a storage class for each probe. This gives extra flexibility without an added configuration burden.

Our use case is not to even expose metrics for storage class we are not interested in (to avoid storage and network).
You can then even tag with storage class even without such filter. That's an evolution.

> Do we need an extra prefixes parameter? Can't you just define a separate target for each prefix?

No

> This makes it difficult to assess each change individually so I'd suggest separate PRs.

I could generally, but for such a small PR, splitting is unnecessary from my point of view.

**Comment by ribbybibby** [comment]
> Our use case is not to even expose metrics for storage class we are not interested in (to avoid storage and network).
You can then even tag with storage class even without such filter. That's an evolution.

How many storage classes are there, 7? In the context of a prometheus exporter, that's not a huge amount of cardinality and I don't think it would add much network burden in terms of the increase in the size of the response. For instance, there's a lot more possible http response codes but in general the ability to drop specific codes is not something that is built into exporters/libraries.

If you want to avoid storing series for particular storage classes, or you only want to keep a particular set, then you could use `metrics_relabel_configs` in Prometheus to achieve that.

For those reasons, I don't think we need filtering of storage classes at the exporter level, as it can be achieved in Prometheus itself.

> No

No we don't need it, or no you can't define a separate target for each prefix? As I see it, you can just enumerate the different bucket/prefix combos as targets so there's no need to add an extra parameter to the s3 exporter. Plus that gives you the benefit of multiple prefixes not sharing the same scrape timeout, or an error for one prefix affecting the scrape of another.

> I could generally, but for such a small PR, splitting is unnecessary from my point of view.

I'm inclined not to accept the `prefixes` parameter or the default options, so I would suggest we make this PR about refining the `storage_class` label (which I think is a very valuable addition).

Thank you by the way for the contribution!

**Comment by cchantep** [comment]
You are free to refactor this. Anyway we use the code as-is in our fork. Best regards.

**Pull request closed without merging**